vllm/model_executor/models/chatglm.py

```diff
@@ -24,11 +28,230 @@
 from vllm.model_executor.layers.vocab_parallel_embedding import (
     ParallelLMHead, VocabParallelEmbedding)
 from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+from vllm.model_executor.models.glm4_vision_encoder import EVA2CLIPModel
 from vllm.model_executor.sampling_metadata import SamplingMetadata
-from vllm.sequence import IntermediateTensors
+from vllm.multimodal.base import MultiModalData
+from vllm.multimodal import MULTIMODAL_REGISTRY, MultiModalDataDict
+from vllm.sequence import (VLLM_TOKEN_ID_ARRAY_TYPE, IntermediateTensors,
+                           SequenceData)
 from vllm.transformers_utils.configs import ChatGLMConfig
 
-from .interfaces import SupportsLoRA
+from .interfaces import SupportsLoRA, SupportsMultiModal
+
+logger = init_logger(__name__)
+
+
+def calculate_image_placeholder(vision_config):
+    return (vision_config["image_size"] // vision_config["patch_size"] // 2)**2
+
+
+# @lru_cache
+def cached_get_image_processor(
+    processor_name: str,
+    *args,
+    trust_remote_code: bool = False,
+    **kwargs,
+):
+    """Gets an image processor for the given model name via HuggingFace."""
+    # don't put this import at the top level
+    # it will call torch.cuda.device_count()
+    from transformers import AutoTokenizer
+
+    try:
+        processor = AutoTokenizer.from_pretrained(
+            processor_name,
+            *args,
+            trust_remote_code=trust_remote_code,
+            **kwargs)
+        image_processor = processor.apply_chat_template
+    except ValueError as e:
+        # If the error pertains to the processor class not existing or not
+        # currently being imported, suggest using the --trust-remote-code flag.
+        # Unlike AutoTokenizer, AutoImageProcessor does not separate such errors
+        if not trust_remote_code:
+            err_msg = (
+                "Failed to load the image processor. If the image processor is "
+                "a custom processor not yet available in the HuggingFace "
+                "transformers library, consider setting "
+                "`trust_remote_code=True` in LLM or using the "
+                "`--trust-remote-code` flag in the CLI.")
+            raise RuntimeError(err_msg) from e
+        else:
+            raise e
+
+    return image_processor
```

Comment:
Since the default `image_processor` method of vllm is not applicable to GLM-4v, I rewrote the `image_processor` method to ensure that the code can work

Reply:
Since `glm-4v` doesn't have `image_processor` implemented and uses tokenizer to process multi-modal data, we can use `cached_get_tokenizer` in [vllm/vllm/multimodal/utils.py](https://github.com/vllm-project/vllm/blob/cdc72e3c80b7029c49de9667150f68481f386956/vllm/multimodal/utils.py#L18):
https://github.com/vllm-project/vllm/blob/cdc72e3c80b7029c49de9667150f68481f386956/vllm/multimodal/utils.py#L18

Reply:
Thanks for your reply, I have modified it in commit 2c899318ee276f97f37814ba8ee471ba9d0d3fd8